### PR TITLE
Add User model and implement login users

### PR DESCRIPTION
**Pull request by idprietop**

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -1,2 +1,12 @@
 class ApplicationController < ActionController::Base
+    helper_method :current_user
+    helper_method :logged_in?
+
+    def current_user
+        User.find_by(id: session[:user_id])
```
Comment (juanhurtado10) [inline app/controllers/application_controller.rb]: @idprietop details matter when you are writing code, please fix this code indentation and make sure you have your editor configured to automatically manage this.

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -1,2 +1,12 @@
 class ApplicationController < ActionController::Base
+    helper_method :current_user
+    helper_method :logged_in?
```
Comment (juanhurtado10) [inline app/controllers/application_controller.rb]: @idprietop what a helper_method is? and what is it used for?

File: app/controllers/articles_controller.rb [hunk 3]
```diff
@@ -1,7 +1,5 @@
 class ArticlesController < ApplicationController
-
-  http_basic_authenticate_with name: "dhh", password: "secret", except: [:index, :show]
-
+  before_action :validate_session
```
Comment (juanhurtado10) [inline app/controllers/articles_controller.rb:2]: @idprietop what a **before_action** means ?

File: app/controllers/articles_controller.rb [hunk 4]
```diff
@@ -30,23 +29,36 @@ def edit
...
+      end
     else
-      render :edit
+      redirect_to @article
     end
   end
 
   def destroy
     @article = Article.find(params[:id])
+    if @article.user_id == current_user.id
```
Comment (juanhurtado10) [inline app/controllers/articles_controller.rb:45]: @idprietop could you please explain this and also fix the code indentation, it is really hard to understand!

File: app/controllers/articles_controller.rb [hunk 5]
```diff
@@ -30,23 +29,36 @@ def edit
...
+      redirect_to @article
+    end
   end
 
   private
-    def article_params
-      params.require(:article).permit(:title, :body, :status)
+
+  def article_params
+    params.require(:article).permit(:title, :body, :status)
```
Comment (juanhurtado10) [inline app/controllers/articles_controller.rb:56]: @idprietop what **require** and **permit** does?

File: app/controllers/sessions_controller.rb [hunk 6]
```diff
@@ -0,0 +1,25 @@
+class SessionsController < ApplicationController
+  before_action :validate_session, only: [:new, :welcome]
+  def create
+    @user = User.find_by(username: params[:username])
+      if @user && @user.authenticate(params[:password])
```
Comment (juanhurtado10) [inline app/controllers/sessions_controller.rb]: @idprietop again, indentation issues

File: app/models/user.rb [hunk 7]
```diff
@@ -0,0 +1,4 @@
+class User < ApplicationRecord
+    has_secure_password
```
Comment (juanhurtado10) [inline app/models/user.rb:2]: @idprietop **has_secure_password** could you please explain this?

File: app/models/user.rb [hunk 8]
```diff
@@ -0,0 +1,4 @@
...
+    has_many :followers, class_name: 'User'
```
Comment (juanhurtado10) [inline app/models/user.rb]: @idprietop why a **class_name** is needed here?

File: app/views/articles/index.html.erb [hunk 9]
```diff
@@ -1,13 +1,19 @@
...
     <li>
-      <%= link_to article.title, article %>
+      <%= link_to article.title, article %> Created by:
+      <%= User.find_by(id: article.user_id).username %>
     </li>
   <% end %>
 </ul>
 
 <%= link_to "New Article", new_article_path %>
+<%= button_to "Logout", logout_path, method: :delete%>
```
Comment (juanhurtado10) [inline app/views/articles/index.html.erb:19]: @idprietop if I am not logged in, I'm gonna see this logout button. Is that ok?

File: app/views/sessions/new.html.erb [hunk 10]
```diff
@@ -0,0 +1,8 @@
+<h1>Login</h1>
+   <%= form_tag '/login' do %>
```
Comment (juanhurtado10) [inline app/views/sessions/new.html.erb]: @idprietop we should not use routes in this way, there are route helper methods to do this. Please correct

File: app/views/sessions/new.html.erb [hunk 11]
```diff
@@ -0,0 +1,8 @@
...
+   <%= label_tag :username%>
+   <%= text_field_tag :username %>
```
Comment (juanhurtado10) [inline app/views/sessions/new.html.erb:4]: @idprietop when you look at **text_field_tag** once the template is rendered, what will you see?

File: app/views/sessions/new.html.erb [hunk 12]
```diff
@@ -0,0 +1,8 @@
...
+   <%= label_tag :password%>
+   <%= password_field_tag :password%>
+   <%= submit_tag "Login"%>
+<%end%>
```
Comment (juanhurtado10) [inline app/views/sessions/new.html.erb]: @idprietop **Tip:** as a good coding practice, all the files should end within a blank line

File: config/routes.rb [hunk 13]
```diff
@@ -1,5 +1,12 @@
 Rails.application.routes.draw do
-  root "articles#index"
+  root 'sessions#welcome'
+
+  get 'login', to: 'sessions#new'
```
Comment (juanhurtado10) [inline config/routes.rb:4]: @idprietop why not just use a resources method to cover all the session actions?

File: db/migrate/20210204144205_create_follows.rb [hunk 14]
```diff
@@ -0,0 +1,10 @@
+class CreateFollows < ActiveRecord::Migration[6.1]
```
Comment (juanhurtado10) [inline db/migrate/20210204144205_create_follows.rb:1]: @idprietop is this related to the login task?

File: db/migrate/20210115203759_create_articles.rb [hunk 15]
```diff
@@ -3,6 +3,7 @@ def change
     create_table :articles do |t|
       t.string :title
       t.text :body
+      t.references :user, null: false, foreign_key: true
```
Comment (juanhurtado10) [inline db/migrate/20210115203759_create_articles.rb:6]: @idprietop is this a good approach? what if I ran this migration already?

**Review by juanhurtado10 — CHANGES_REQUESTED** [review]
@idprietop This is a good PR however there is some work that needs to be done.

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (idprietop) [reply app/controllers/application_controller.rb]: I will make sure that the editor is configured to manage the code automatically and I will take into account the details when writing the code.

File: app/controllers/application_controller.rb (hunk 2 above)
Comment (idprietop) [reply app/controllers/application_controller.rb]: helper_mehod is a helper method which allows us to explicitly share some methods defined in the controller so that they are available in the view.

File: app/controllers/articles_controller.rb (hunk 3 above)
Comment (idprietop) [reply app/controllers/articles_controller.rb:2]: before_action means that before executing any action in the controller, what is defined in the before_action will be executed first.

File: app/controllers/articles_controller.rb (hunk 4 above)
Comment (idprietop) [reply app/controllers/articles_controller.rb:45]: In this line of code I am making sure that when a user is going to delete an article, first validate if this article belongs to the user who logged in.

File: app/controllers/articles_controller.rb (hunk 5 above)
Comment (idprietop) [reply app/controllers/articles_controller.rb:56]: Requires an instance of an article and permit to receive the parameters: title, body and status.

File: app/views/sessions/new.html.erb (hunk 11 above)
Comment (idprietop) [reply app/views/sessions/new.html.erb:4]: Once the template is rendered the text_fiel_tag creates a standard text field and is used to enter smaller pieces of text.

File: app/views/sessions/new.html.erb (hunk 12 above)
Comment (idprietop) [reply app/views/sessions/new.html.erb]: Thanks for the tip, I'll keep it in mind.

File: config/routes.rb (hunk 13 above)
Comment (idprietop) [reply config/routes.rb:4]: You're right it's a better idea, I didn't do it because I applied it the same way it was in the documentation.

File: db/migrate/20210204144205_create_follows.rb (hunk 14 above)
Comment (idprietop) [reply db/migrate/20210204144205_create_follows.rb:1]: It is not related to that task, surely I added it before uploading the PR and forgot to remove it.

File: app/models/user.rb (hunk 8 above)
Comment (idprietop) [reply app/models/user.rb]: I also forgot to remove this, this line is not part of this PR.

File: app/views/articles/index.html.erb (hunk 9 above)
Comment (idprietop) [reply app/views/articles/index.html.erb:19]: If you have not logged in, you will not see the button because the before_action validates the session and if you have not logged in it redirects to the login.

File: app/models/user.rb (hunk 7 above)
Comment (idprietop) [reply app/models/user.rb:2]: has_secure_password is used to encrypt and authenticate passwords using BCrypt. It assume that the model has a column named password_digest which will be encrypted by BCrypt.

File: db/migrate/20210115203759_create_articles.rb (hunk 15 above)
Comment (idprietop) [reply db/migrate/20210115203759_create_articles.rb:6]: It is not a good approach, ideally the migration should not be edited later to maintain data integrity.

File: app/controllers/application_controller.rb (hunk 2 above)
Comment (juanhurtado10) [reply app/controllers/application_controller.rb]: @idprietop good, what makes different a helper method defined in the Application controller?

File: app/controllers/articles_controller.rb (hunk 4 above)
Comment (juanhurtado10) [reply app/controllers/articles_controller.rb:45]: @idprietop ok, good!

File: app/models/user.rb (hunk 7 above)
Comment (juanhurtado10) [reply app/models/user.rb:2]: @idprietop excellent!

File: app/models/user.rb (hunk 8 above)
Comment (juanhurtado10) [reply app/models/user.rb]: ok!

File: db/migrate/20210204144205_create_follows.rb (hunk 14 above)
Comment (juanhurtado10) [reply db/migrate/20210204144205_create_follows.rb:1]: ok! Thanks!

File: db/migrate/20210115203759_create_articles.rb (hunk 15 above)
Comment (juanhurtado10) [reply db/migrate/20210115203759_create_articles.rb:6]: @idprietop that's correct! did you fix this?

File: app/controllers/articles_controller.rb (hunk 5 above)
Comment (juanhurtado10) [reply app/controllers/articles_controller.rb:56]: @idprietop thanks!

File: app/views/articles/index.html.erb (hunk 9 above)
Comment (juanhurtado10) [reply app/views/articles/index.html.erb:19]: @idprietop that is the answer that I'm looking for! 👏🏻

File: db/migrate/20210204144205_create_follows.rb (hunk 14 above)
Comment (juanhurtado10) [inline db/migrate/20210204144205_create_follows.rb:1]: @idprietop if this is not related to this PR, please remove it, thank you!

File: app/controllers/application_controller.rb (hunk 2 above)
Comment (idprietop) [reply app/controllers/application_controller.rb]: What makes it different to define helper methods in the application controller is that the other controllers will have access to these methods because they inherit from this controller so it will not be necessary to define them in each of the controllers.

File: db/migrate/20210204144205_create_follows.rb (hunk 14 above)
Comment (idprietop) [reply db/migrate/20210204144205_create_follows.rb:1]: Ok, I already removed it.

File: app/views/sessions/new.html.erb (hunk 10 above)
Comment (idprietop) [reply app/views/sessions/new.html.erb]: Ok, I already made the change.
